src/vm/security.rs

```diff
@@ -1,99 +1,88 @@
-///   - There mustn't be accesses to the program segment outside the program
+///   - There must not be accesses to the program segment outside the program
 ///     data range.
+///   - All addresses in memory must be real (not temporary)
 ///
 /// Note: Each builtin is responsible for checking its own segments' data.
 pub fn verify_secure_runner(
     runner: &CairoRunner,
     verify_builtins: bool,
     vm: &mut VirtualMachine,
 ) -> Result<(), VirtualMachineError> {
-    let program_base = runner
-        .program_base
-        .as_ref()
-        .ok_or(RunnerError::NoProgBase)?;
-
-    let builtin_segments = match verify_builtins {
+    let builtins_segment_info = match verify_builtins {
         true => runner.get_builtin_segments_info(vm)?,
         false => HashMap::new(),
     };
-
-    let builtin_segment_by_index = builtin_segments
-        .iter()
-        .map(|(seg_name, seg_info)| (seg_info.index, (seg_name, seg_info)))
-        .collect::<HashMap<_, _>>();
-
-    let memory_iter = vm
+    // Check builtin segment out of bounds.
+    for (index, stop_ptr) in builtins_segment_info {
+        let current_size = index
+            .to_usize()
+            .and_then(|index| vm.memory.data.get(index))
+            .map(|segment| segment.len());
+        // + 1 here accounts for maximum segment offset being segment.len() -1
+        if current_size >= Some(stop_ptr + 1) {
+            return Err(VirtualMachineError::OutOfBoundsBuiltinSegmentAccess);
+        }
+    }
+    // Check out of bounds for program segment.
+    let program_segment_index = runner
+        .program_base
+        .and_then(|rel| rel.segment_index.to_usize())
+        .ok_or(RunnerError::NoProgBase)?;
+    let program_segment_size = vm
         .memory
         .data
-        .iter()
-        .enumerate()
-        .flat_map(|(idx, segment)| {
-            segment.iter().enumerate().filter_map(move |(off, value)| {
-                value
-                    .as_ref()
-                    .map(|val| (Relocatable::from((idx as _, off)), val))
-            })
-        });
-    for (addr, value) in memory_iter {
-        // Check builtin segment bounds.
-        if let Some((_, seg_info)) = builtin_segment_by_index.get(&addr.segment_index) {
-            if addr.offset >= seg_info.size {
-                return Err(RunnerError::FailedMemoryGet(MemoryError::NumOutOfBounds).into());
+        .get(program_segment_index)
+        .map(|segment| segment.len());
+    // + 1 here accounts for maximum segment offset being segment.len() -1
+    if program_segment_size >= Some(runner.program.data.len() + 1) {
+        return Err(VirtualMachineError::OutOfBoundsProgramSegmentAccess);
+    }
+    // Check that the addresses in memory are valid
+    // This means that every temporary address has been properly relocated to a real address
+    // Asumption: If temporary memory is empty, this means no temporary memory addresses were generated and all addresses in memory are real
```

Comment:
What if the address was stored but not accessed? Then (I think) it wouldn't create any values in `temp_data`, nor would it fail, but it would still store an address to a temporary segment. Is this reasoning right?

Reply:
`RelocatableValue` and `MaybeRelocatable` are not exposed during hints, neither in the original nor in our cairo_rs_py version, so there is no way to "create" a relocatable value. The only way to obtain a relocatable value with negative index is to create a temporary segment, and when this happens, a new segment will be created in the temp_data, leading to it being not empty.
I made this assumption as having temporary data is not something that happens when running normal cairo programs, and this function is ran after every standard (aka no special flags) run, so being able to forgo iterating the whole memory will present an advantage performace-wise.